internal/postgres/password/scram.go

```diff
@@ -0,0 +1,225 @@
+package password
+
+/*
+ Copyright 2020 Crunchy Data Solutions, Inc.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import (
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"hash"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/xdg/stringprep"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+// scramVerifierFormat is the format of the value that is stored by PostgreSQL
+// and follows the format:
+//
+// <DIGEST>$<ITERATIONS>:<SALT>$<STORED_KEY>:<SERVER_KEY>
+//
+// where:
+// DIGEST = SCRAM-SHA-256 (only value for now in PostgreSQL)
+// ITERATIONS = the number of iteratiosn to use for PBKDF2
+// SALT = the salt used as part of the PBKDF2, stored in base64
+// STORED_KEY = the hash of the client key, stored in base64
+// SERVER_KEY = the hash of the server key
+const scramVerifierFormat = "SCRAM-SHA-256$%d:%s$%s:%s"
+
+// These constants are defined as part of the PostgreSQL implementation for
+// SCRAM, but can be overriden by the user
+// https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/include/common/scram-common.h
+const (
+	// scramDefaultIterations is the number of iterations to make as part of the PBKDF2
+	// algorithm
+	scramDefaultIterations = 4096
+	// scramDefaultSaltLength is the length of the generated salt used in creating the
+	// hashed password
+	scramDefaultSaltLength = 16
+)
+
+// scramDefaultHash is the hashing algorithm to use
+var scramDefaultHash = sha256.New
+
+// the following are used as part of the SCRAM verifier generation
+var (
+	scramClientKeyMessage = []byte("Client Key")
+	scramServerKeyMessage = []byte("Server Key")
+)
+
+var (
+	// ErrSCRAMPasswordInvalid is returned when the password attributes are invalid
+	ErrSCRAMPasswordInvalid = errors.New(`invalid password attributes. must provide "password"`)
+	// ErrSCRAMSaltLengthInvalid is returned when the salt length is less than 1
+	ErrSCRAMSaltLengthInvalid = errors.New(`salt length must be at least 1`)
+)
+
+// SCRAMPassword contains the building blocks to build a PostgreSQL SCRAM
+// verifier. Implements the PostgresPassword interface
+type SCRAMPassword struct {
+	// Hash is the hash algorithm to use. Currently the only thing supported in
+	// PostgreSQL is SHA-256
+	Hash func() hash.Hash
```

Comment:
This looks unused. `SCRAMPassword.hash` and `SCRAMPassword.hmac` take an argument now.

Reply:
I think this was an attempt to "future proof" for new hashing methods, but in the end, it did not work out all too well. Perhaps in the future when it's needed.